### ISSUE TITLE
Request: clean gradle cache from time to time...

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [gradle/gradle#28263]
### Issue type

Polishing

### Problem description

I was told to write about this here:
https://issuetracker.google.com/issues/326453909#comment2

**Comment by ov7a** [comment]
We need more information to understand the request. Could you please provide more context about what you want to achieve?

If we receive a clear use case description, we may be able to suggest a workaround.

This issue will be closed after 7 days unless you provide more information.

***

Can you describe in more details what problem you are facing? Please, share all relevant in formation in the issue instead of pasting a link.
See also this [documentation](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home:cache_cleanup).

**Comment by AndroidDeveloperLB** [comment]
@ov7a Clear case of what?
I've already written that it takes huge amounts of space.
For some users , it reached 125GB ...

**Comment by ov7a** [comment]
@AndroidDeveloperLB What is the list of the files? What are their sizes? What version of gradle you are using? etc.

**Comment by AndroidDeveloperLB** [comment]
@ov7a Notice I wrote "for some users". I mean it's not me. I mean others.
You can ask anything you wish the users that have it on the link I've provided. They have it much more than me.
As for me, I already removed folders there. Now it takes much less.

**Comment by ov7a** [comment]
Sorry, this issue is not actionable then. I'm gonna close it.
Our available resources are severely limited, and we must be sure we are looking at the exact problem users are facing, not just vague reports.

**Comment by AndroidDeveloperLB** [comment]
@ov7a What's "vogue" about it? I already said I had it too. For me it took  37.5 GB!
Does it seem normal?
Should it only increase its usage over time?
There is no limit to how much it takes? Even for things that it doesn't use for a very long time?

**Comment by ov7a** [comment]
@AndroidDeveloperLB The problem is that random numbers and claims are not actionable. 
As I said, our resources are limited to investigate every complaint.
I gave you the link to the documentation about cache clean-up. 

If you believe you found a bug, then please, report it properly, so it can be actionable for us to resolve.

Please note that this repo is subject to [CoC](https://gradle.org/conduct/), and It is probably better to stay constructive.

**Comment by AndroidDeveloperLB** [comment]
@ov7a It's a cache folder. Of course the numbers would be random, as each has his own experience of which dependencies he comes across. 
The resources aren't related. The issue exists. 
How can I report it "properly"? It's a cache folder...
It accumulates over a long time. Months even. I can't record a video that takes months. That's impractical.
What you need is to check the cleaning mechanism of the folder, and if there isn't one, add it. It's a request to handle it properly. To avoid letting it to reach huge sizes.

**Comment by ov7a** [comment]
I've already suggested what you can [provide](https://github.com/gradle/gradle/issues/28263#issuecomment-1971219042).

**Comment by AndroidDeveloperLB** [comment]
@ov7a You want me to give you information of what I have now, or before?
If now, please explain how to check each. Is there some command to run? Where to search for them?

**Comment by ov7a** [comment]
> You want me to give you information of what I have now, or before?
At the moment when you think the cache size is too large

>  If now, please explain how to check each. Is there some command to run? Where to search for them?
You can use `du -h` for that.

> detailed sizes
Is seems like the upload was interrupted, please re-upload the file

**Comment by AndroidDeveloperLB** [comment]
@ov7a The gradle command got me this:

```
Welcome to Gradle 8.2!

Here are the highlights of this release:
 - Kotlin DSL: new reference documentation, assignment syntax by default
 - Kotlin DSL is now the default with Gradle init
 - Improved suggestions to resolve errors in console output

For more details see https://docs.gradle.org/8.2/release-notes.html


------------------------------------------------------------
Gradle 8.2
------------------------------------------------------------

Build time:   2023-06-30 18:02:30 UTC
Revision:     5f4a070a62a31a17438ac998c2b849f4f6892877

Kotlin:       1.8.20
Groovy:       3.0.17
Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023
JVM:          21 (Oracle Corporation 21+35-LTS-2513)
OS:           Windows 11 10.0 amd64
```

About the du command, I don't get how to make it show anything useful.

I tried what you wrote, and it just showed me this:

```
DU v1.62 - Directory disk usage reporter
Copyright (C) 2005-2018 Mark Russinovich
Sysinternals - www.sysinternals.com
```

So I tried to write the path to the folder (also tried with `"`) :

```
./du -h C:\Users\User\.gradle
```

And it just showed me the manual of it.

**Comment by ov7a** [comment]
@AndroidDeveloperLB did you read the manual for `du`?
Please try `du -c -v C:\Users\User\.gradle`

**Comment by AndroidDeveloperLB** [comment]
@ov7a I used what you wanted, with "-h" alone. Only tried to fix by setting the path, as it seemed missing. 
Now what you want is very different, and it produces a lot of output so I had to put it into a file.

Attached here.

[output.zip](https://github.com/gradle/gradle/files/14459679/output.zip)

**Comment by ov7a** [comment]
Thank you for providing the data.

From the file you provided:
1. The total usage is 35.6 GB
2. 15.9 GB are used by the daemon folders. This is related to logs, and it's a known issue:
* https://github.com/gradle/gradle/issues/2688
3. The next biggest folder is `caches` with 16,2 GB of data
The largest entries are `build-cache-1`, `jars-9`, `wrapper/dists`. All of them are subject to retention clean-up as documented [here](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home:cache_cleanup). If you think they are not cleaned up as expected, please provide evidence for that.

See also:
* https://github.com/gradle/gradle/issues/27426

**Comment by AndroidDeveloperLB** [comment]
@ov7a How can I show evidence of this now? You want to see if there are files in these folders that are older than 30 days?
Please explain what to check.

**Comment by ov7a** [comment]
> You want to see if there are files in these folders that are older than 30 days?
Yes, please provide a list of files with creation and access times.

**Comment by AndroidDeveloperLB** [comment]
@ov7a So in "C:\Users\User\.gradle\caches", the screenshot I've provided already shows there are folders there that are quite old:
https://github.com/gradle/gradle/issues/28263#issuecomment-1971311672

Here it is again, sorted by time:

![image](https://github.com/gradle/gradle/assets/5357526/4647c275-836f-4a68-a656-c0664d36fb3e)

Oldest folders are of 2021-2023...

**Comment by ov7a** [comment]
Please, provide the data in text format _for files_ including both creation and access times.

**Comment by AndroidDeveloperLB** [comment]
@ov7a Please be more specific. Are you talking about files within the folders? If so which, and what command or window do you want me to show?
Or are you talking about the folders as I've shown here?

**Comment by ov7a** [comment]
> Are you talking about files within the folders?

Yes

> what command 

You can use [`forfiles`](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/forfiles) for that.

**Comment by AndroidDeveloperLB** [comment]
@ov7a You are still not specific, so I chose myself, and you tell me if it's ok for you.
I tried "forfiles", and it doesn't show dates at all. It's worse than the simple "dir". Please explain what is the command you want to execute. 

For now, instead I wrote: 
```
dir C:\Users\User\.gradle\caches\journal-1
```

Output:

    Directory: C:\Users\User\.gradle\caches\journal-1

```
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----          3/4/2024   5:34 PM        7413534 file-access.bin
-a----        10/13/2021  12:23 PM             65 file-access.properties
-a----          3/4/2024   5:56 PM             39 journal-1.lock
```

I also tried to use the app "File locator pro", and just sorted by time created.
It seems there are about 305,441 files&folders.

I've found there an example file of 2021 called "file-access.properties" in "C:\Users\User\.gradle\caches\journal-1", as both its creation and modification dates are on 2021. 


![image](https://github.com/gradle/gradle/assets/5357526/dd847a75-be1b-4ca0-bc0e-77a35a5b5724)

My guess is that you want to find the folder that its content is all old (both creation&modification time, and of all files and not specific ones), because then it means it's truly old.
But I don't know how to perform this check for you.

**Comment by ov7a** [comment]
Did you read the docs for `forfiles`? It mentions `@FDATE` and allows you to execute any command on the file. Sorry, please figure it out yourself or search for help on SO and other places. I'm trying to help you with Gradle, not with how Windows works and which commands is has to do basic operations.

> My guess is that you want to find the folder that its content is all old (both creation&modification time, and of all files and not specific ones), because then it means it's truly old.

You are correct.

**Comment by AndroidDeveloperLB** [comment]
@ov7a No, I already asked you for the specific command and you only gave me a link to the command. 
As you want a very specific command, please write it down. I can't know what you want to do.

**Comment by ov7a** [comment]
> As you want a very specific command

I don't care about the command you execute. I need the data to be able to help you and triage the issue. I specified the request a few messages back:

> Please, provide the data in text format for files including both creation and access times.

I tried to help you by pointing out the sample command and the documentation. Note that I said "you can use", not "you must use". I don't think I need to be 100% specific and give you the exact instructions on how to extract the data I asked for: it's a basic file operation. Again, if you're struggling with that - please, search for help on other resources: this knowledge is not in any way specific to Gradle.

**Comment by AndroidDeveloperLB** [comment]
@ov7a Sorry but I never had to handle such a specific task for files. This isn't a part of my knowledge either. What I use usually is simple apps as I've mentioned, and even then it's not for such a task.
If you want me to write you the output of some command, you need to be specific about the command , and not telling me to read manuals.
I don't need help. I said that if you want specific output, show me exactly what you want me to execute, because I don't know which output will be fine for you anyway.
I showed you the output as I've found it. Isn't it enough?

**Comment by ov7a** [comment]
In the current state, I don't see this issue as actionable.

If you don't care about it, let's just leave it closed.

if you do care about it, then please improve your bug report, as I stated before.

I tried to be helpful, but I don't feel that you really want to invest your time into having a proper report and somebody fixing the issue.

**Comment by AndroidDeveloperLB** [comment]
@ov7a Sorry but I don't know what you want. I've already found a folder that seems to be old and showed you that it has both old and new files.
My guess is that it doesn't prove anything, because some were accessed recently, so gradle probably found it useful to let it stay, no matter when it was created. If that's the case, this folder isn't a good sample, and you need to tell me where you want to check next. 
Isn't this good enough for you?
It's not practical for me to go over thousands of files manually and check it out.... Also not practical to go over many manuals without even knowing what you want to have printed, and I don't even know if the manuals can help you and find a folder that has all of its files to be old.

It's a very customized query that you probably want to have, right?
You probably want to find a folder that all of its files were not accessed in a long time, doesn't matter when it or its files were created.
I'm not an expert at all in commands lines for Windows for such a customized request. 
So I tell you again: If you know of a specific command to run, write it.
Otherwise I can't help you with this.

The folders are "random". Each user will have a different experience. 
I suggest to talk with the person who had the largest storage being used. On the link I've provided, it's the creator of the post, probably.

**Comment by mavek87** [comment]
@AndroidDeveloperLB du is a linux command. It should print a list of your files (a very long list). Install in your windows machine the linux subsystem for Windows. Here are the doc => https://learn.microsoft.com/en-US/windows/wsl/install 

after that you enter in the terminal using the linux subsystem and execute du pointing your .gradle folder in your home.

something like **du -h .gradle**

probably you need to save the output into a file you could try to do something like **du -h .gradle > file-gradle.txt**

Search on internet how to install the linux subsystem and how it works du in linux. Unfortunately windows doesn't have anything like this to print the content of a folder into a file. Or at least I don't know if something like this exists.
